taichi/codegen/codegen_llvm.cpp

```diff
@@ -1069,14 +1073,56 @@ void CodeGenLLVM::visit(GlobalStoreStmt *stmt) {
   TI_ASSERT(!stmt->parent->mask() || stmt->width() == 1);
   TI_ASSERT(llvm_val[stmt->data]);
   TI_ASSERT(llvm_val[stmt->ptr]);
-  builder->CreateStore(llvm_val[stmt->data], llvm_val[stmt->ptr]);
+  if (auto cit = stmt->ptr->ret_type.ptr_removed()->cast<CustomIntType>()) {
+    // 1. load byte pointer
+    auto byte_ptr_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(0)});
+    auto byte_ptr = builder->CreateLoad(byte_ptr_in_bit_struct);
+    assert(byte_ptr->getType()->getPointerElementType()->isIntegerTy(8));
+
+    // 2. load bit offset
+    auto bit_offset_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(1)});
+    auto bit_offset = builder->CreateLoad(bit_offset_in_bit_struct);
+    assert(bit_offset->getType()->isIntegerTy(32));
+
+    builder->CreateCall(get_runtime_function("set_partial_bits_b32"),
+                        {builder->CreateBitCast(byte_ptr, llvm::Type::getInt32PtrTy(*llvm_context))
+                             , bit_offset, tlctx->get_constant(cit->get_num_bits()), llvm_val[stmt->data]});
+  }
+  else {
+    builder->CreateStore(llvm_val[stmt->data], llvm_val[stmt->ptr]);
+  }
 }
 
 void CodeGenLLVM::visit(GlobalLoadStmt *stmt) {
   int width = stmt->width();
   TI_ASSERT(width == 1);
-  llvm_val[stmt] = builder->CreateLoad(tlctx->get_data_type(stmt->ret_type),
-                                       llvm_val[stmt->ptr]);
+  TI_INFO("stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer(): {}", stmt->ptr->ret_type->cast<PointerType>()->is_bit_pointer());
+  if (auto cit = stmt->ret_type->cast<CustomIntType>()) {
+    TI_INFO("bit-level global load {}", cit->to_string());
+    // 1. load byte pointer
+    auto byte_ptr_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(0)});
+    auto byte_ptr = builder->CreateLoad(byte_ptr_in_bit_struct);
+    assert(byte_ptr->getType()->getPointerElementType()->isIntegerTy(8));
+    auto bit_level_container = builder->CreateLoad(builder->CreateBitCast(byte_ptr, llvm::Type::getInt32PtrTy(*llvm_context)));
+
+    // 2. load bit offset
+    auto bit_offset_in_bit_struct = builder->CreateGEP(llvm_val[stmt->ptr], {tlctx->get_constant(0), tlctx->get_constant(1)});
+    auto bit_offset = builder->CreateLoad(bit_offset_in_bit_struct);
+    assert(bit_offset->getType()->isIntegerTy(32));
```

Comment:
I will move these repeating lines(same in GlobalStoreStmt and GlobalLoad) into a new function.